<a href="https://colab.research.google.com/github/dhatri-12/DSA/blob/main/covtypeEC.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from sklearn.datasets import fetch_covtype
from sklearn import tree
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

# Load dataset into ctype
ctype=fetch_covtype(shuffle=True)



### About the Datatset:

Dataset name : Covertype Dataset

This data set is used for classifying pixels into one of seven forest cover types.

The dataset contains 581,012 examples.

It features 54 attributes, including both categorical and integer types.

The attributes encompass various environmental factors such as elevation, aspect, slope, and distances to hydrological features.

There are also binary columns representing wilderness areas and soil types. The classes correspond to forest cover types.

In [2]:
X, y = ctype.data, ctype.target

clf = tree.DecisionTreeClassifier()

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=0)

y_pred_50 = clf.fit(X_train, y_train).predict(X_test)
print("Number of mislabeled points Decision Tree out of a total %d points : %d"
       % (X_test.shape[0], (y_test != y_pred_50).sum()))

accuracy_dt_50 = accuracy_score(y_test, y_pred_50)
print(f"Decision Tree Accuracy for 50% test data: {accuracy_dt_50}")


Number of mislabeled points Decision Tree out of a total 290506 points : 22589
Decision Tree Accuracy for 50% test data: 0.9222425698608635


In [3]:
clf = tree.DecisionTreeClassifier()

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

y_pred_20 = clf.fit(X_train, y_train).predict(X_test)

print("Number of mislabeled points Decision Tree out of a total %d points : %d"
       % (X_test.shape[0], (y_test != y_pred_20).sum()))

accuracy_dt_20 = accuracy_score(y_test, y_pred_20)
print(f"Decision Tree Accuracy for 20% test data: {accuracy_dt_20}")

Number of mislabeled points Decision Tree out of a total 116203 points : 6970
Decision Tree Accuracy for 20% test data: 0.9400187602729706


In [4]:
# split data to train/test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=0)

#Single nearest neighbor
neigh = KNeighborsClassifier(n_neighbors=1)

#Build NN classifier and predict
sy_pred = neigh.fit(X_train, y_train).predict(X_test)

print("Number of mislabeled points 1NN out of a total %d points : %d"
       % (X_test.shape[0], (y_test != sy_pred).sum()))

accuracy_nn = accuracy_score(y_test, sy_pred)

print(f"Nearest Neighbor Accuracy: {accuracy_nn}")

Number of mislabeled points 1NN out of a total 290506 points : 11406
Nearest Neighbor Accuracy: 0.9607374718594452


### **Decision Tree Classifier Performance**

Decision Tree accuracy for 50% test data is 92.22%.

Decision Tree accuracy for 20% test data is 94.00%.

This improvement between the two test sizes (50% vs. 20%) accuracy is about 1.78%.

### **k-Nearest Neighbors Classifier (1-NN) Performance**

When using 50% of the dataset for testing, the 1-NN classifier hits an accuracy of 96.07%.


### **Comparing the two models:**


The 1-NN model is more accurate than the Decision Tree when both use half the data for testing. It got 96.07% right compared to 92.22%.

But, 1-NN takes more time and computer power, especially with big datasets. The Decision Tree is quicker and easier to understand because it shows you how it makes decisions.